daemon: Sync local IPs to lxcmap periodically

jrajahalme opened the pull request:
LXCMap should not get out of sync, but there is some evidence that
sometimes it does. Add a new controller to refresh the host entries in
the lxcmap every 5 seconds, but only if they are not already there.

No garbage collection of potentially stale host entries in the lxcmap
is done.

lxcmap implementation of bpf.MapValue is fixed to use a pointer receiver
so that lookups function as intended.

Fixes: #3870

joestringer commented on daemon/daemon.go:
```diff
@@ -905,6 +883,51 @@ func (d *Daemon) init() error {
+
+						localIPIDPair := identity.IPIdentityPair{IP: ip, ID: identity.ReservedIdentityHost}
+						d.OnIPIdentityCacheChange(ipcache.Upsert, localIPIDPair)
+						envoy.NetworkPolicyHostsCache.OnIPIdentityCacheChange(ipcache.Upsert, localIPIDPair)
```
I believe that this will cause problems if we ever do this insert twice.

EDIT: Which *shouldn't* happen if the `ipcache.IPIdentityCache.LookupByIP()` logic works as expected. Should be fine, I think.

joestringer commented on daemon/daemon.go:
```diff
@@ -905,6 +883,51 @@ func (d *Daemon) init() error {
+			DoFunc: func() error {
+				// TODO: Update addresses first, in case node addressing has changed.
+				// TODO: Once these start changing on runtime, figure out the locking strategy.
+				// TODO: Delete stale host entries from the lxcmap.
```
If you're planning to bind lxcmap update to ipcache update, then ipcache update already has GC, so maybe you could put the GC there..? (See `OnIPIdentityCacheGC()`).

jrajahalme replied:
`OnIPIdentityCacheGC()` deletes stale entries from ipcache bpf map. Here we are adding potentially missing entries. I agree that these should be combined into a more comprehensive desired/realized state implementation, but that is out of scope for this PR.

joestringer reviewed (commented):
It's not obvious in your commit message, but this is performing ipcache sync as well. This has some implications. I would suggest running this with short timers and L7 policies, and keep an eye on things like the logs to see that Cilium is operating sanely.

joestringer commented on pkg/maps/lxcmap/lxcmap.go:
```diff
@@ -192,10 +193,20 @@ func WriteEndpoint(f EndpointFrontend) error {
+	key := NewEndpointKey(ip)
+	value, err := LXCMap.Lookup(key)
+	if err != nil || value.(*EndpointInfo).Flags&EndpointFlagHost == 0 {
+		log.WithError(err).WithField(logfields.IPAddr, ip).Debugf("Adding local ip to endpoint map")
```
There's nothing that is specific about this being a `local ip` here, perhaps we can shift this to the caller.

tgraf replied:
I would prefer to return an error here and log in the caller

raybejjani replied:
I third this.

jrajahalme replied:
Done.

tgraf commented on pkg/bpf/endpoint.go:
```diff
@@ -40,7 +40,7 @@ type EndpointKey struct {
 func (k EndpointKey) GetKeyPtr() unsafe.Pointer { return unsafe.Pointer(&k) }
 
-// GetValuePtr returns the unsafe pointer to the BPF value
+// GetValuePtr returns the unsafe pointer to the BPF KEY <<<< NOTE!
```
Please move this commit to a separate PR. It needs improvements but I think fixing the comment like this is not what we want because it was still not clear to me even after this comment ;-)

Let's move it to a separate PR as we won't have to backport this.

raybejjani replied:
lbmap has a similar bug, if you end up making a separate PR to fix this.

jrajahalme replied:
@raybejjani Next commit fixes the value receiver for lbmap, but I just noted that I did not fix it for EndpointKey on which the wrong comment was on here. So the comment is wrong, and this too returns a pointer to the temporary. Fixing this forces all other methods on EndpointKey to also use pointer receiver, which is correct as well, as what is a point of taking a copy and then returning pointer to it. That fix will likely cascade further than we want for the bug fix, so I'll not touch it for now.

tgraf commented on daemon/daemon.go:
```diff
@@ -905,6 +883,51 @@ func (d *Daemon) init() error {
+			},
+			RunInterval: time.Duration(5) * time.Second,
+		})
+	// TODO: To keep the old behavior we should wait for one successful completion of the controller
```
We can't keep this as a todo as we might get into a situation where no host IP are synced down.

You can do the following:
- extract the controller run function into a separate function and run it once and fail on any error to keep behaviour on failing while bootstrapping when any of the host IPs can't be added.
- or create channel and pass it into the controller and then close it on the first successful run. You could then wait on the channel to be closed outside with a timeout. This would give us retries even on bootstrapping but the logic is a bit more complicated.

jrajahalme replied:
I'll do the first, as it is less complicated.

tgraf reviewed (changes requested):
Please submit patch 3 separately as it fixes a separate bug as well. We are doing backports on PR level so for fixes that we target for backports, we should try to isolate PRs as much as possible.

raybejjani commented on daemon/daemon.go:
```diff
@@ -905,6 +883,51 @@ func (d *Daemon) init() error {
+					id, exists := ipcache.IPIdentityCache.LookupByIP(ip.String())
+					if !exists || id != identity.ReservedIdentityHost {
+						// Upsert will not propagate (reserved:host->ID) mappings across the cluster.
+						// Manually push it into each IPIdentityMappingListener.
```
@joestringer has a PR to add a UpdateToKVStore. It does propagate to the listeners via the KVStore. If we use that this could then be:
```golang
ipcache.IPIdentityCache.Upsert(ip.String(), identity.ReservedIdentityHost)
ipcache.IPIdentityCache.UpsertToKVStore(ip.String(), identity.ReservedIdentityHost) <-- I don't remember the actual parameters
```

Upsert takes a local lock so the new info is available immediately, UpsertToKVStore would call the listeners eventually. I don't know how sensitive we are to delays, though.

joestringer replied:
The reserved identities should specifically *not* be upserted to the kvstore. Host is semantically only the host namespace where the pod is running, not other nodes in the cluster.

raybejjani replied:
Oops. You are most correct. I'm still a little uneasy about invoking the listeners directly but this is the same init-time one-off we had so nothing really changed, I guess.

raybejjani commented on daemon/daemon.go:
```diff
@@ -905,6 +883,51 @@ func (d *Daemon) init() error {
 
+// startLXCMapSync spawns a controller which synchronizes the host entries in BPF lxc Map
+// with the in-memory host IP address values.
+func (d *Daemon) startLXCMapSync() error {
```
Since this is a bit of a hack-fix, could you mention that here? Ideally we can ditch the controller one day.

jrajahalme replied:
Will add a comment.

raybejjani reviewed (approved):
the 1-line change to endpoint is approved. The other comments are just suggestions

jrajahalme reviewed (commented):
@tgraf lookups on the lxcmap depend on the GetValuePtr to have a pointer receiver, so there is a dependency between the commits. Do you mean that I should split this up to 3 separate PRs and then we backport the latter 2, in the order of dependency?

jrajahalme commented on pkg/maps/lxcmap/lxcmap.go:
```diff
@@ -192,10 +193,20 @@ func WriteEndpoint(f EndpointFrontend) error {
 	return LXCMap.Update(key, ep)
 }
 
+func SyncHostEntry(ip net.IP) error {
```
Fixed.

joestringer commented on daemon/daemon.go:
```diff
@@ -905,6 +899,43 @@ func (d *Daemon) init() error {
+// hosts cache if changes were made.
+func (d *Daemon) syncLXCMap() error {
+	// TODO: Update addresses first, in case node addressing has changed.
+	// TODO: Once these start changing on runtime, figure out the locking strategy.
```
What specifically are we deferring figuring out here? Locking of `lxcmap`? of `IPIdentityCache`?

jrajahalme replied:
All of it, I guess.

joestringer commented on daemon/daemon.go:
```diff
@@ -905,6 +899,43 @@ func (d *Daemon) init() error {
+// ipcache, if needed, and also notifies the daemon and network policy
+// hosts cache if changes were made.
+func (d *Daemon) syncLXCMap() error {
+	// TODO: Update addresses first, in case node addressing has changed.
```
What is this todo and what are the implications of not doing it now?

jrajahalme replied:
If host device configuration (e.g., IP addressing) changes we do not reflect that change to the bpf maps, i.e., things will break.

joestringer reviewed (approved):
Seems fine to me, so long as the todos aren't pressing.

jrajahalme committed:
maps: Use pointer receivers for MapValue types.

bpf.MapValue interface function GetValuePtr() returns a pointer to a
new temporary if the function receiver is a value rather than a
pointer.

endpoint, lxcmap, ipcache, and lbmap were also using value receivers
for their implementations of MapValue interface. The problem with this
is that any lookups would fail to return the actual value, as the
bpf.LookupElement would write the value into a temporary unaccessible
to the caller. No such lookups were performed, so this did not cause
any problems in practice. Fix the implementations to prevent future
problems.

This fix is otherwise low risk, but it has happened earlier in
development that GetValuePtr() implmentations were not fixed properly
and a pointer to the pointer receiver was returned. This is not
noticed by the compiler, and would result in garbage data being
written to/read from the bpf maps.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
daemon: Sync local IPs to lxcmap periodically.

LXCMap should not get out of sync, but there is some evidence that
sometimes it does. Add a new controller to refresh the host entries in
the lxcmap every 5 seconds, but only if they are not already there.

No garbage collection of potentially stale host entries in the lxcmap
is done.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
